src/secret_connection.rs

```diff
@@ -0,0 +1,452 @@
+    // https://github.com/tendermint/tendermint/blob/013b9cef642f875634c614019ab13b17570778ad/p2p/conn/secret_connection.go#L208-L238
+    let mut remote_eph_pubkey = vec![];
+    merge(
+        WireType::LengthDelimited,
+        &mut remote_eph_pubkey,
+        &mut amino_buf,
+    ).unwrap();
+    let mut remote_eph_pubkey_fixed: [u8; 32] = Default::default();
+    remote_eph_pubkey_fixed.copy_from_slice(&remote_eph_pubkey[..32]);
+
+    Ok(remote_eph_pubkey_fixed)
+}
+
+// Returns recv secret, send secret, challenge as 32 byte arrays
+fn derive_secrets_and_challenge(
+    shared_secret: &[u8; 32],
+    loc_is_lo: bool,
+) -> ([u8; 32], [u8; 32], [u8; 32]) {
+    let info = b"TENDERMINT_SECRET_CONNECTION_KEY_AND_CHALLENGE_GEN";
+    let hk = Hkdf::<Sha256>::extract(None, shared_secret);
+    let hkdf_vector = hk.expand(&*info, 96);
+
+    let challenge_vector = &hkdf_vector[64..96];
+    let mut challenge: [u8; 32] = [0; 32];
+    challenge.copy_from_slice(challenge_vector);
+    let mut recv_secret = [0u8; 32];
+    let mut send_secret = [0u8; 32];
+    if loc_is_lo {
+        recv_secret.copy_from_slice(&hkdf_vector[0..32]);
+        send_secret.copy_from_slice(&hkdf_vector[32..64]);
+    } else {
+        send_secret.copy_from_slice(&hkdf_vector[0..32]);
+        recv_secret.copy_from_slice(&hkdf_vector[32..64]);
+    }
+    (recv_secret, send_secret, challenge)
+}
+
+// Return is of the form lo, hi
+fn sort32(first: [u8; 32], second: [u8; 32]) -> ([u8; 32], [u8; 32]) {
+    if second > first {
+        (first, second)
+    } else {
+        (second, first)
+    }
+}
+
+// Sign the challenge with the local private key
+fn sign_challenge(challenge: [u8; 32], local_privkey: &DalekSigner) -> Result<Signature, Error> {
+    local_privkey
+        .sign(&challenge)
+        .map_err(|e| err!(SigningError, "{}", e))
+}
+
+#[derive(Clone, PartialEq, Message)]
+struct AuthSigMessage {
+    #[prost(bytes, tag = "1")]
+    key: Vec<u8>,
+    #[prost(bytes, tag = "2")]
+    sig: Vec<u8>,
+}
+
+// TODO(ismail): change from DecodeError to something more generic
+// this can also fail while writing / sending
+fn share_auth_signature<IoHandler: io::Read + io::Write + Send + Sync>(
+    sc: &mut SecretConnection<IoHandler>,
+    pubkey: &[u8; 32],
+    signature: Signature,
+) -> Result<AuthSigMessage, DecodeError> {
+    let amsg = AuthSigMessage {
+        key: pubkey.to_vec(),
+        sig: signature.into_bytes().to_vec(),
+    };
+    let mut buf: Vec<u8> = vec![];
+    amsg.encode(&mut buf).unwrap();
+
+    sc.write_all(&buf).unwrap();
+
+    let mut rbuf = vec![0; 100]; // 100 = 32 + 64 + (amino overhead)
+    sc.read_exact(&mut rbuf).unwrap();
+
+    // TODO: proper error handling:
+    Ok(AuthSigMessage::decode(&rbuf)?)
+}
+
+// TODO: Check if internal representation is big or small endian
+// increment nonce big-endian by 2 with wraparound.
+fn incr_nonce(nonce: &mut [u8; 12]) {
+    for i in (0..AEAD_NONCE_SIZE).rev() {
+        nonce[i] += 1;
```

Comment:
```
---- secret_connection::tests::test_incr_nonce stdout ----
	thread 'secret_connection::tests::test_incr_nonce' panicked at 'attempt to add with overflow', src/secret_connection.rs:356:9
stack backtrace:
```

I think this would be a lot clearer (and panic-free!) if you just kept an internal 64-bit 
counter and used [BigEndian::write_u64](https://docs.rs/byteorder/1.2.4/byteorder/trait.ByteOrder.html#tymethod.write_u64) to serialize it.

That'd also make it easier to switch to little endian, if there's still time to do that :wink: